utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/encode/AvroSchemaOutputValidator.scala

```diff
@@ -0,0 +1,292 @@
+      val fieldPath = buildPath(key, path)
+      validateTypingResult(value, schema.getValueType, fieldPath)
+    }.foldLeft[ValidatedNel[OutputValidatorError, Unit]](().validNel)((a, b) => a combine b)
+
+    schemaFieldsValidationResult
+  }
+
+  private def validateArraySchema(array: TypedClass, schema: Schema, path: Option[String]) = {
+     val valuesValidationResult: ValidatedNel[OutputValidatorError, Unit] = array
+        .params
+        .zipWithIndex
+        .map { case (el, index) =>
+          val pathKey = array.params match {
+            case _ :: Nil => "" //One element list is 'general' object, and we preset it as field[]
+            case _ => index.toString
+          }
+
+          val elementPath = buildPath(pathKey, path, isGeneric = true)
+          validateTypingResult(el, schema.getElementType, elementPath)
+        }
+        .foldLeft[ValidatedNel[OutputValidatorError, Unit]](().validNel)((a, b) => a combine b)
+
+    valuesValidationResult
+  }
+
+  private def validateUnionSchema(union: TypingResult, schema: Schema, path: Option[String]) = {
+    implicit class UnionValidationResult(results: List[ValidatedNel[OutputValidatorError, Unit]]) {
+      type ValidatedType = ValidatedNel[OutputValidatorError, Unit]
+      def toValidated: ValidatedNel[OutputValidatorError, Unit] =
+        if (results.exists(_.isValid)) valid else results.foldLeft[ValidatedType](().validNel)((a, b) => a combine b)
+    }
+
+    object UnionNullableSchemaType {
+      private val MaxSizeUnion = 2
+      def unapply(schema: Schema): Option[Schema] = {
+        Option(schema).filter(_.getTypes.size() == MaxSizeUnion).filter(_.isNullable).flatMap(_.getTypes.asScala.find(!_.isNullable))
+      }
+    }
+
+    //check is there only one typing error with exactly same field as path - it means there was checking whole object (without going deeper e.g. List/Map/Record)
+    def singleObjectTypingError(errors: NonEmptyList[OutputValidatorError]): Boolean =
+      errors.collect{case err: OutputValidatorTypeError => err} match {
+        case head :: Nil => path.exists(_.equals(head.field))
+        case _ => false
+      }
+
+    def createUnionValidationResults(withoutNullability: Boolean): List[ValidatedNel[OutputValidatorError, Unit]] = {
+      val schemas = if(withoutNullability) schema.getTypes.asScala.filterNot(_.isNullable) else schema.getTypes.asScala
+      schemas.map(validateTypingResult(union, _, path)).toList
+    }
+
+    val results = schema match {
+      case UnionNullableSchemaType(_) => //Nullability: union[null, any] it's most common situation, so we want to handle that in special way
+        val notNullableResults = createUnionValidationResults(withoutNullability = true)
+        notNullableResults.toValidated match {
+          case Invalid(errors) if singleObjectTypingError(errors) => //when single typing error is true, we have to validate again including nullability
```

Comment:
1. Why UnionNullableSchemaType is special case?
2. Why u use extractor if you don't use extracted schema?
3. Why we have to "validate again including nullability"?
4. double negation is hard - I mean by "including nullability" you do withoutNullability = false

Reply:
1. I get it
2. Now is fine
3. I get it
4. Now is better, but IMO we still can work a bit on this name to make this hard part more understable. You don't check nullability because you already know that this schema is nullable. You check if provided type matches at lease one element of union of schemas. So maybe it should be "includeNullTypeInChecking" or just "checkNullSchemaType"?